ntoskrnl/io/iomgr/iomgr.c

```diff
@@ -552,15 +569,30 @@ IoInitSystem(IN PLOADER_PARAMETER_BLOCK LoaderBlock)
 
     /* Load boot start drivers */
     IopInitializeBootDrivers();
+    if (!IopWaitForBootDevicesStarted())
+    {
+        DPRINT1("IopWaitForBootDevicesStarted failed!\n");
+        return FALSE;
+    }
 
     /* Call back drivers that asked for */
     IopReinitializeBootDrivers();
+    if (!IopWaitForBootDevicesStarted())
+    {
+        DPRINT1("IopWaitForBootDevicesStarted failed!\n");
+        return FALSE;
+    }
 
     /* Check if this was a ramdisk boot */
     if (!_strnicmp(LoaderBlock->ArcBootDeviceName, "ramdisk(0)", 10))
     {
         /* Initialize the ramdisk driver */
         IopStartRamdisk(LoaderBlock);
+        if (!IopWaitForBootDevicesStarted())
```

Comment:
It should be noted somewhere that if that three changes are understandable, these are hacks and should be performed in the called functions which should return only when the devices are started.

Reply:
So you don't like the placement or the whole way how it is done? 
I think the latest two statements can be changed to one placed after if(...ramdisk(0)...).
But

Reply:
The wait code shouldn't be in `IoInitSystem()` but it should be in the called functions themselves.
That means the functions should only return when done.